integration/container/health_test.go

```diff
@@ -0,0 +1,48 @@
+package container
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/docker/docker/api/types"
+	"github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/api/types/network"
+	"github.com/docker/docker/api/types/strslice"
+	"github.com/docker/docker/integration/util/request"
+	"github.com/stretchr/testify/require"
+)
+
+// TestHealthCheckWorkdir verifies that health-checks inherit the containers'
+// working-dir.
+func TestHealthCheckWorkdir(t *testing.T) {
+	defer setupTest(t)()
+	ctx := context.Background()
+	client := request.NewAPIClient(t)
+
+	c, err := client.ContainerCreate(ctx,
+		&container.Config{
+			Image:      "busybox",
+			Tty:        true,
+			WorkingDir: "/foo",
+			Cmd:        strslice.StrSlice([]string{"top"}),
+			Healthcheck: &container.HealthConfig{
+				Test:     []string{"CMD-SHELL", "if [ \"$PWD\" = \"/foo\" ]; then exit 0; else exit 1; fi;"},
+				Interval: 500 * time.Millisecond,
+				Retries:  1,
+			},
+		},
+		&container.HostConfig{},
+		&network.NetworkingConfig{},
+		"healthtest",
+	)
+	require.NoError(t, err)
+	err = client.ContainerStart(ctx, c.ID, types.ContainerStartOptions{})
+	require.NoError(t, err)
+
+	time.Sleep(1 * time.Second)
```

Comment:
I'm not sure if an integration test is necessary here. Could this be tested with a unit  test of `run()` instead?

Sleeping in a test case should generally be avoided. If this integration test is really necessary then this should use `poll.WaitOn()` to wait for the state, instead of sleep and inspect.

`poll.WaitOn()` will also make the test more reliable on loaded systems, because the timeout can be longer than 1s.

Reply:
I think an integration test is warranted here, because `run` and `exec` (used for the healthcheck) take different code paths.

Thanks for the `poll.WaitOn()`, I'll check that one out